### ISSUE TITLE
Deadlock on exception throw

### ISSUE THREAD

**Issue by bradphelan** [taskflow/taskflow#479]
**Describe the bug**
If we wait on a taskflow that throws an exception then the executor deadlocks on wait(). I would have expected at least terminate to be called in this instance.

**To Reproduce**
```cpp
tf::Taskflow tf;
tf.emplace(
	[]{ throw std::runtime_error("x");}
);
auto e = std::make_shared<tf::Executor>(1);
e->run(tf).wait();
```

**Desktop (please complete the following information):**
```
Edition	Windows 11 Pro
Version	22H2
Installed on	‎25/‎01/‎2023
OS build	22621.1555
Experience	Windows Feature Experience Pack 1000.22640.1000.0
Device name	baron-papanoida
Processor	AMD Ryzen 9 5900X 12-Core Processor               3.70 GHz
Installed RAM	64.0 GB
Device ID	35EC4264-F7B8-42BF-9AAD-0B4079DF952A
Product ID	00330-80000-00000-AA067
System type	64-bit operating system, x64-based processor
Pen and touch	No pen or touch input is available for this display

Microsoft Visual Studio Professional 2019
Version 16.11.20
VisualStudio.16.Release/16.11.20+32929.386
Microsoft .NET Framework
Version 4.8.09032

Installed Version: Professional

Visual C++ 2019   00435-30013-98631-AA953
Microsoft Visual C++ 2019

ASP.NET and Web Tools 2019   16.11.110.1451
ASP.NET and Web Tools 2019

Azure App Service Tools v3.0.0   16.11.110.1451
Azure App Service Tools v3.0.0

C# Tools   3.11.0-4.22108.8+d9bef045c4362fbcab27ef35daec4e95c8ff47e1
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

GitExtensions   1.0
Git Extensions is a graphical user interface for Git that allows you to control Git without using the command-line

Graphical Debugging Extension   1.0
Graphical Debugging Visual Studio Extension allows to view graphical representation of variables, e.g. Boost.Geometry models or vectors of values.

Include Toolbox 2019   3.2.67
Various tools for managing C/C++ #includes: Formatting, sorting, exploring, pruning. VS2019 compatibility version. For VS2022 visit https://marketplace.visualstudio.com/items?itemName=Agrael.IncludeToolbox2022

IncrediBuild Build Acceleration   1.5.0.29
IncrediBuild effectively reduces compilation and development times by up to 90%.

Intel® VTune™ Profiler 2023   1.3.3
Intel® VTune™ Profiler 2023, (build 1.3.3),  Copyright © 2009-2022 Intel Corporation. All rights reserved.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

NuGet Package Manager   5.11.2
NuGet Package Manager in Visual Studio. For more information about NuGet, visit https://docs.nuget.org/

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

Snippet Designer   1.8.1
Snippet Designer is a Visual Studio plug in which allows you to create and search for snippets inside the IDE

Test Adapter for Boost.Test   1.0
Enables Visual Studio's testing tools with unit tests written for Boost.Test.  The use terms and Third Party Notices are available in the extension installation directory.

Test Adapter for Google Test   1.0
Enables Visual Studio's testing tools with unit tests written for Google Test.  The use terms and Third Party Notices are available in the extension installation directory.

TypeScript Tools   16.0.30526.2002
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   3.11.0-4.22108.8+d9bef045c4362fbcab27ef35daec4e95c8ff47e1
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Tools for CMake   1.0
Visual Studio Tools for CMake

VsVim   2.8.0.0
VsVim is a Vim emulator for Visual Studio
```

**Comment by bradphelan** [comment]
I notice that

![image](https://user-images.githubusercontent.com/17650/235195871-283238b1-f17d-407c-87a1-fab3de103f78.png)

is never reached if the task throws an exception. It seems that the teardown of the topology just never happens. Shouldn't the structure of task dependencies be exception safe somehow.

**Comment by bradphelan** [comment]
It seems taskflow doesn't really stick to the spirit of what users of std::future would expect. See here for a test case

![image](https://user-images.githubusercontent.com/17650/235202654-c8bc02da-0a54-4aee-b31d-4f0e832ef868.png)

and the output is

![image](https://user-images.githubusercontent.com/17650/235202828-765c269d-a6d8-4a0d-b30d-cf4892f7d585.png)

This shows that the exception is propagated from the async code. *faulty* is indeed run on the threadpool as can be seen by this stack trace.

![image](https://user-images.githubusercontent.com/17650/235203115-0dcb0ea9-e135-46b9-be41-7f7f53757510.png)

In the unit test it can be seen that the exception with text *X* is thrown on when the user calls *wait* on the future.

Given that task flow is using std::future to handle waiting on taskflow then it would seem that it should handle exceptions in an expected way.

**Comment by bradphelan** [comment]
I notice you have https://github.com/taskflow/taskflow/issues/193 where you say you won't do this. Can you please add some information on why.

**Comment by bradphelan** [comment]
I also note that TBB by intel supports exception propagation. 
https://www.inf.ed.ac.uk/teaching/courses/ppls/TBBtutorial.pdf
![image](https://user-images.githubusercontent.com/17650/235871830-2c269b5e-7b56-4eca-9605-3245e0d2141a.png)

**Comment by tsung-wei-huang** [comment]
Hi @bradphelan , currently, we ask applications to explicitly catch exceptions inside their tasks, instead of propagating them because running a taskflow will have many simultaneously running tasks. We will need to have a lock and other global control to store potential many exceptions that may be thrown by multiple tasks. At this moment, I do not have a clear idea about what a good design would be. Any thoughts?

**Comment by bradphelan** [comment]
I think you need to make this somehow optional via a protocol on the executor. 

- catch no exceptions - terminate on first uncaught
- catch all exceptions and forward - wait for all tasks to finish and collect all exceptions
- catch one exception - wait for all tasks to finish or abort all tasks via a cancellation token on first exception found

**Comment by danielytics** [comment]
Note that exceptions are already caught within a worker and passed to [WorkerInterface::scheduler_epilogue](https://taskflow.github.io/taskflow/classtf_1_1WorkerInterface.html#a3e6d68fd4041f433d1b7ca9e5786b57c), the problem is that if the executor is waiting, it will deadlock.

So I suggest the following solution: _"catch one exception per worker"_. If an exception is thrown, shut the worker down (calling scheduler_epilogue in the process, if one is configured) and cancel the execution as you would if the following code were run:

```
tf::Future fu = executor.run(taskflow);
fu.cancel();
fu.get();
```
Any tasks that are in-flight can still be completed on their particular worker and each worker can handle up to one exception (but since no new tasks will be scheduled, that's ok).

**Comment by tsung-wei-huang** [comment]
@bradphelan @danielytics the exception support is up in the current master/dev branch. Examples can be found [here](https://taskflow.github.io/taskflow/ExceptionHandling.html). Please share any feedback. This support will be part of release v3.7. 

Thank you for your patience.

**Comment by danielytics** [comment]
@tsung-wei-huang That looks good!

The exception_ptr probably isn’t needed in `tf::WorkerInterface::scheduler_epilogue` anymore, then, since I don’t think there’s any way an exception can be received there since task exceptions are now handled in this way.

**Comment by tsung-wei-huang** [comment]
@danielytics yes, you are right. I will remove that `tf::WorkerInterface::scheduler_epilogue` from the next release.

**Comment by danielytics** [comment]
You can keep the epilogue, but remove the parameter.

**Comment by danielytics** [comment]
@tsung-wei-huang I see you've removed `WorkerInterface` altogether, but I still rely on it to set thread names and affinity :fearful:
